### PR TITLE
Improving performance

### PR DESCRIPTION
**Pull request by matteofabbri**
Major changes: 
- Added caching on Type Parser 
- Optimized linq queries 

Minor changes: removed unaccessed field, accessors, variables

### CODE REVIEW

File: OrleansDashboard.Client/DashboardClient.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        private readonly IGrainFactory grainFactory;
+
+        public DashboardClient(IGrainFactory grainFactory)
+        {
+            this.grainFactory = grainFactory;
+        }
+
+        public async Task<Immutable<DashboardCounters>> DashboardCounters()
+        {
+            var grain = grainFactory.GetGrain<IDashboardGrain>(0);
```
Comment (SebastianStehle) [inline OrleansDashboard.Client/DashboardClient.cs]: You can store this grain as a field.

File: OrleansDashboard.Client/IDashboardClient.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+using Orleans.Concurrency;
+using Orleans.Runtime;
+using OrleansDashboard.Client.Model;
+using OrleansDashboard.Client.Model.History;
+
+namespace OrleansDashboard.Client
+{
+    public interface IDashboardClient
+    {
+        Task<Immutable<DashboardCounters>> DashboardCounters();
```
Comment (SebastianStehle) [inline OrleansDashboard.Client/IDashboardClient.cs]: I would abstract the immutability away.

File: OrleansDashboard/Metrics/TypeFormatting/TypeFormatter.cs [hunk 3]
```diff
@@ -1,56 +1,32 @@
...
     public class TypeFormatter
     {
-        enum ParseState
-        {
-            TypeNameSection,
-            GenericCount,
-            GenericArray,
-            TypeArray
-        }
+        static MemoryCache cache = MemoryCache.Default;
```
Comment (SebastianStehle) [inline OrleansDashboard/Metrics/TypeFormatting/TypeFormatter.cs]: I would use a concurrent dictionary. There is no need to ever remove something from the cache as the number of entries is relatively small.

File: OrleansDashboard/DashboardTelemetryConsumer.cs [hunk 4]
```diff
@@ -87,40 +88,32 @@ public void Flush()
...
-            foreach (var metric in timespanMetrics.ToArray())
-            {
-                var v = metric.Value.Current;
-                var d = metric.Value.Current - metric.Value.Last;
-
-                counters.Add(new StatCounter { Name = metric.Key, Value = v.ToString("c", CultureInfo.InvariantCulture), Delta = d.ToString("c", CultureInfo.InvariantCulture) });
-            }
-
-            if (counters.Count > 0)
+            var countersArray = metrics.Select(metric => new StatCounter
```
Comment (SebastianStehle) [inline OrleansDashboard/DashboardTelemetryConsumer.cs]: Actually there is no performance improvement here. In Addition: You have a view more allocations.

Comment (SebastianStehle) [reply OrleansDashboard/DashboardTelemetryConsumer.cs]: You could just change the signature of the grain from array to list

File: OrleansDashboard/Metrics/Details/SiloStatusOracleSiloDetailsProvider.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+﻿using System.Linq;
```
Comment (SebastianStehle) [inline OrleansDashboard/Metrics/Details/SiloStatusOracleSiloDetailsProvider.cs]: We should track this change as a rename in git

**Comment by SebastianStehle** [comment]
My reviews are just some comments, I think the proposed changes do not really matter (on the other hand the performance improvements are only small too ;))

File: OrleansDashboard.Client/DashboardClient.cs (hunk 1 above)
Comment (matteofabbri) [reply OrleansDashboard.Client/DashboardClient.cs]: right :) fixed

File: OrleansDashboard/Metrics/TypeFormatting/TypeFormatter.cs (hunk 3 above)
Comment (matteofabbri) [reply OrleansDashboard/Metrics/TypeFormatting/TypeFormatter.cs]: right :) fixed

File: OrleansDashboard/DashboardTelemetryConsumer.cs (hunk 4 above)
Comment (matteofabbri) [reply OrleansDashboard/DashboardTelemetryConsumer.cs]: "there is no performance improvement here" 
Well, if you think that two .ToArray() don't affect performance....

"You have a view more allocations"
And where ?

Comment (richorama) [reply OrleansDashboard/DashboardTelemetryConsumer.cs]: We have a benchmarking project in the solution, which is great for testing perf assumptions, and evidencing improvements (we've had performance problems reported in the past).

**Comment by richorama** [comment]
Thanks both :¬)

**Pull request merged**